mitiq/zne/scaling/folding.py

```diff
@@ -421,6 +289,262 @@ def fold_all(
+            weight_mask=[1.0, 0.5, 2.0, 0.0],
+            scale_factor=4,
+            folding_method="from_left",
+        )
+        [2, 2, 1, 0]
+    """
+
+    if not 1.0 <= scale_factor:
+        raise ValueError(
+            f"Requires scale_factor >= 1 but scale_factor = {scale_factor}."
+        )
+
+    # Find the maximum odd integer smaller or equal to scale_factor
+    num_uniform_folds = int((scale_factor - 1.0) / 2.0)
+    odd_integer_scale_factor = 2 * num_uniform_folds + 1
+
+    # Uniformly folding all gates to reach odd_integer_scale_factor
+    num_folds_mask = []
+    for w in weight_mask:
+        if np.isclose(w, 0.0):
+            num_folds_mask.append(0)
+        else:
+            num_folds_mask.append(num_uniform_folds)
+
+    # If the scale_factor is an odd integer, we are done.
+    if np.isclose(odd_integer_scale_factor, scale_factor):
+
+        return num_folds_mask
+
+    # If necessary, fold a subset of gates to approximate the scale_factor
+    input_circuit_weight = sum(weight_mask)
+    output_circuit_weight = odd_integer_scale_factor * input_circuit_weight
+
+    # Express folding order through a list of indices
+    folding_order = list(range(len(weight_mask)))
+    if folding_method == "from_left":
+        pass
+    elif folding_method == "from_right":
+        folding_order.reverse()
+    elif folding_method == "at_random":
+        rnd_state = np.random.RandomState(seed)
+        rnd_state.shuffle(folding_order)
+    else:
+        raise ValueError(
+            "The option 'folding_method' is not valid."
+            "It must be 'at_random', or 'from_left', or 'from_right'."
+        )
+
+    # Fold gates until the input scale_factor is better approximated
+    approx_error = np.abs(
+        output_circuit_weight - scale_factor * input_circuit_weight
+    )
+    for j in folding_order:
+        # Skip gates with 0 weight
+        if np.isclose(weight_mask[j], 0.0):
+            continue
+
+        # Compute the approx error if a new fold would be applied
+        new_output_circuit_weight = output_circuit_weight + 2 * weight_mask[j]
+        new_approx_error = np.abs(
+            new_output_circuit_weight - scale_factor * input_circuit_weight
+        )
+        # Fold the candidate gate only if it helps improving the approximation
+        if new_approx_error < approx_error:
+            approx_error = new_approx_error
+            output_circuit_weight = new_output_circuit_weight
+            num_folds_mask[j] += 1
+        # TODO: decide if removing next else case to get better approximations
+        else:
+            break
+
+    return num_folds_mask
+
+
+def _apply_fold_mask(
+    circuit: Circuit,
+    num_folds_mask: List[int],
+    squash_moments: Optional[bool] = True,
+) -> Circuit:
+    r"""Applies local unitary folding to the gates of the input circuit
+    according to the input num_folds_mask.
+
+    More precicely, G_j^\dag G_j is applied after the j_th gate G_j of
```

Comment:
Same (minor) comment above about the linear indexing a circuit.

Reply:
e063dfa